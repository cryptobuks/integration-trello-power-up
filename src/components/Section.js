import React from 'react'
import SectionItem from './SectionItem.js'
import update from 'react-addons-update'
import { isInVisionUrl } from '../util/url.js'
import T from '../util/trello.js'
import auth from '../util/auth.js'

export default class Section extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      attachments: []
    }
    this.fetch = this.fetch.bind(this)
    this.t = T.iframe()
    auth.init()
  }

  componentDidMount () {
    this.t.render(this.fetch)
  }

  componentDidUpdate () {
    return this.t.sizeTo('#root')
  }

  fetch () {
    return this.t.card('all').then(c => {
      let cover = c.cover
      let a = c.attachments.filter(a => isInVisionUrl(a.url))
      let newState = update(this.state, {
        $set: {
          attachments: a,
          cover: cover,
          card: c
        }
      })
      this.setState(newState)
    })
  }

  cover (attachment) {
    if (this.state.cover && attachment.url === this.state.cover.name) {
      return this.state.cover
    }
  }

  render () {
    return (
      <div className='u-clearfix'>
        {this.state.attachments.map(attachment =>
          <SectionItem
            {...this.props}
            attachment={attachment}
            card={this.state.card}
            cover={this.cover(attachment)}
            key={attachment.id}
            updateCover={this.fetch}
          />
        )}
      </div>
    )
  }
}
