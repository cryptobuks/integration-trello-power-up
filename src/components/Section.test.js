import React from 'react'
import { mount } from 'enzyme'
import Promise from 'bluebird'
import sinon from 'sinon'
import toJson from 'enzyme-to-json'

import T from '../util/trello.js'
// import attachment from '../util/attachment.js'

import Section from './Section'
// import SectionItem from './SectionItem.js'

jest.mock('../util/trello.js')
jest.mock('../util/config.js')
jest.mock('../util/auth.js')
jest.mock('../util/cover.js')
jest.mock('../util/attachment.js')

const setup = card => {
  const ctx = {}
  T.iframe().card.mockReturnValue(Promise.resolve(card))
  ctx.wrapper = mount(<Section />, {
    disableLifecycleMethods: false
  })

  ctx.component = ctx.wrapper.instance()
  expect(ctx.component).toBeDefined()
  expect(Section.prototype.componentDidMount.callCount).toBe(1)
  expect(ctx.component.t.render).toBeCalledWith(ctx.component.fetch)
  expect(toJson(ctx.wrapper)).toMatchSnapshot()
  return ctx
}

describe('Section', () => {
  beforeAll(() => {
    sinon.spy(Section.prototype, 'componentDidMount')
    sinon.spy(Section.prototype, 'setState')
  })

  beforeEach(() => {
    Section.prototype.componentDidMount.reset()
    Section.prototype.setState.reset()
  })

  it('should render items', () => {
    const ctx = setup({
      attachments: [
        { id: 1, url: 'https://invis.io/NQ8979O8R' },
        { id: 2, url: 'https://google.com' },
        { id: 3, url: 'https://invis.io/JHLK234K2' },
        { id: 4, url: 'https://invision.com' },
        { id: 5, url: 'https://invis.io/LKASHFL23' }
      ],
      cover: {
        name: 'https://invis.io/JHLK234K2'
      }
    })
    return Promise.attempt(ctx.component.fetch).then(() => {
      expect(ctx.component.state.attachments).toEqual(
        expect.arrayContaining([
          { id: 1, url: 'https://invis.io/NQ8979O8R' },
          { id: 3, url: 'https://invis.io/JHLK234K2' },
          { id: 5, url: 'https://invis.io/LKASHFL23' }
        ])
      )
      expect(ctx.component.setState.callCount).toBe(1)
      expect(T.iframe().sizeTo).toBeCalledWith('#root')
      // expect(ctx.wrapper.find(SectionItem)).to.have.length(3)
    })
  })

  it('does not render items', () => {
    const ctx = setup({ attachments: [] })

    return Promise.attempt(ctx.component.fetch).then(() => {
      ctx.component.fetch()
      expect(ctx.component.setState.callCount).toBe(1)
      expect(ctx.component.state.attachments).toEqual(
        expect.arrayContaining([])
      )
      expect(ctx.component.setState.callCount).toBe(1)
      expect(T.iframe().sizeTo).toBeCalledWith('#root')
      // expect(ctx.wrapper.find(SectionItem)).to.have.length(0)
    })
  })

  it('should get cover for attachment', () => {
    const component = new Section({})
    component.state = {
      cover: {
        name: 'https://invis.io/JHLK234K2'
      }
    }
    expect(
      component.cover({ url: 'https://invis.io/NQ8979O8R' })
    ).toBeUndefined()

    expect(component.cover({ url: 'https://invis.io/JHLK234K2' })).toBe(
      component.state.cover
    )
  })
})
