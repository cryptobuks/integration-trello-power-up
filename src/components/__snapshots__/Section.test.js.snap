// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section does not render items 1`] = `
<Section>
  <div
    className="u-clearfix"
  />
</Section>
`;

exports[`Section should render items 1`] = `
<Section>
  <div
    className="u-clearfix"
  />
</Section>
`;
